src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+import ForwardDiff
+import MacroTools
+import LinearAlgebra
+
+# TODO allow providing a pair of address namespaces, one for the model and one
+# for the proposal, when calling another involution function with call
+
+# TODO proposal_retval is a way that continuous data can leak unaccounted for..
+# for now, the requirement is that the return value of f_disc cannot depend on
+# continuous addresses in the model or proposal although in the future, we
+# could do AD through its return value as well using choice_gradients() and add
+# these to the Jacobian...?
+
+# TODO add more checks that we don't visit the same address twice (these can be
+# only enabled in 'check' mode)
+
+struct InvolutionDSLProgram
+    fn!::Function
+end
+
+struct FirstPassState
+    trace
+    u::ChoiceMap
+    constraints::ChoiceMap
+    u_back::ChoiceMap
+    t_cont_reads::Dict
+    u_cont_reads::Dict
+    t_cont_writes::Dict
+    u_cont_writes::Dict
+    t_copy_reads::DynamicSelection
+    u_copy_reads::DynamicSelection
+    marked_as_retained::Set
```

Comment:
brief docstring about what "retained" means (could be "see such-and-such place for more info") might help here